File: pico/bench/symbolize.py
#!/usr/bin/env python3
"""Turn a captured profile into a list of function names.

    ./run.py <image>.uf2 --out capture.txt
    ./symbolize.py ../../build-pico/O2/yax86_pico_bench.elf capture.txt

The firmware prints buckets as addresses because carrying a symbol table on
the board would cost more flash than the profiler is worth. Each bucket is
attributed to whichever function contains its base address.

Read the top few entries and check anything below them before acting on it.
The tail is not reliable: a bucket lands on the function that starts before
it, so a small function inlined into a larger one is credited to whatever
symbol happens to precede it.
"""

import bisect
import re
import subprocess
import sys

BUCKET_RE = re.compile(r"^p ([0-9a-f]{8}) (\d+)$")
TOP_N = 26


def load_symbols(elf):
    """Function symbols from the ELF, sorted by address, as (addr, size, name)."""
    out = subprocess.run(
        ["arm-none-eabi-nm", "-nS", elf], capture_output=True, text=True).stdout
    syms = []
    for line in out.splitlines():
        parts = line.split()
        # nm prints size only for symbols that have one.
        if len(parts) == 4:
            addr, size, kind, name = parts
        elif len(parts) == 3:
            addr, kind, name = parts
            size = "0"
        else:
            continue
        if kind.lower() in "tw":
            syms.append((int(addr, 16), int(size, 16), name))
    syms.sort()
    return syms


def main():
    if len(sys.argv) != 3:
        raise SystemExit("usage: symbolize.py <elf> <capture>")
    elf, capture = sys.argv[1], sys.argv[2]
    syms = load_symbols(elf)
    starts = [s[0] for s in syms]

    def lookup(address):
        i = bisect.bisect_right(starts, address) - 1
        if i < 0:
            return "?"
        start, size, name = syms[i]
        # Past the end of the symbol it follows, so it belongs to no function
        # this ELF names - padding, or a section the profiler covers but nm
        # does not describe.
        if size and address >= start + size:
            return "(gap after %s)" % name
        return name

    totals, total = {}, 0
    with open(capture) as f:
        for line in f:
            match = BUCKET_RE.match(line.strip())
            if not match:
                continue
            name = lookup(int(match.group(1), 16))
            count = int(match.group(2))
            totals[name] = totals.get(name, 0) + count
            total += count
    if total == 0:
        raise SystemExit("no profile buckets in %s" % capture)

    print("  %-46s %8s %7s" % ("symbol", "samples", "share"))
    for name, count in sorted(totals.items(), key=lambda kv: -kv[1])[:TOP_N]:
        print("  %-46s %8d %6.1f%%" % (name[:46], count, 100.0 * count / total))
    print("\n  attributed samples: %d" % total)


if __name__ == "__main__":
    main()

File: pico/bench/symbolize.js
#!/usr/bin/env node

const fs = require('fs/promises');
const {spawnSync} = require('child_process');

// Turns a captured profile into a list of function names.
//
//   ./run.js <image>.uf2 --out capture.txt
//   ./symbolize.js ../../build-pico/O3/yax86_pico_bench.elf capture.txt
//
// The firmware prints buckets as addresses because carrying a symbol table on
// the board would cost more flash than the profiler is worth. Each bucket is
// attributed to whichever function contains its base address.
//
// Read the top few entries and check anything below them before acting on it.
// The tail is not reliable: a bucket lands on the function that starts before
// it, so where several small functions share a bucket only the first is named.

// A bucket line, as the firmware prints it: an address and a sample count.
const kBucketPattern = /^p ([0-9a-f]{8}) (\d+)$/;
// How many symbols to print. Beyond this the shares are too small to mean
// anything, for the reason above.
const kTopSymbols = 26;
// nm's type letters for code: text, and weak symbols that resolve to it.
const kCodeSymbolTypes = 'tw';

// Function symbols from the ELF, sorted by address.
function loadSymbols(elfFilePath) {
  const nm = spawnSync('arm-none-eabi-nm', ['-nS', elfFilePath], {
    encoding: 'utf8',
  });
  if (nm.status !== 0) {
    throw new Error(`arm-none-eabi-nm failed on ${elfFilePath}`);
  }
  const symbols = [];
  for (const line of nm.stdout.split('\n')) {
    const parts = line.split(/\s+/).filter((part) => part.length > 0);
    // nm prints a size only for the symbols that have one.
    let address, size, type, name;
    if (parts.length === 4) {
      [address, size, type, name] = parts;
    } else if (parts.length === 3) {
      [address, type, name] = parts;
      size = '0';
    } else {
      continue;
    }
    if (!kCodeSymbolTypes.includes(type.toLowerCase())) {
      continue;
    }
    symbols.push({
      address: parseInt(address, 16),
      size: parseInt(size, 16),
      name,
    });
  }
  symbols.sort((a, b) => a.address - b.address);
  return symbols;
}

// The last symbol starting at or before an address, by binary search.
function findSymbol(symbols, address) {
  let low = 0;
  let high = symbols.length;
  while (low < high) {
    const middle = (low + high) >> 1;
    if (symbols[middle].address <= address) {
      low = middle + 1;
    } else {
      high = middle;
    }
  }
  return low > 0 ? symbols[low - 1] : null;
}

function symbolNameAt(symbols, address) {
  const symbol = findSymbol(symbols, address);
  if (symbol === null) {
    return '?';
  }
  // Past the end of the symbol it follows, so it belongs to no function this
  // ELF names - padding, or a section the profiler covers but nm does not
  // describe.
  if (symbol.size > 0 && address >= symbol.address + symbol.size) {
    return `(gap after ${symbol.name})`;
  }
  return symbol.name;
}

async function symbolize(elfFilePath, captureFilePath) {
  const symbols = loadSymbols(elfFilePath);
  const samplesByName = new Map();
  let total = 0;
  const capture = await fs.readFile(captureFilePath, 'utf8');
  for (const line of capture.split('\n')) {
    const match = kBucketPattern.exec(line.trim());
    if (match === null) {
      continue;
    }
    const name = symbolNameAt(symbols, parseInt(match[1], 16));
    const count = Number(match[2]);
    samplesByName.set(name, (samplesByName.get(name) ?? 0) + count);
    total += count;
  }
  if (total === 0) {
    throw new Error(`no profile buckets in ${captureFilePath}`);
  }

  console.log(
    '  ' + 'symbol'.padEnd(46) + 'samples'.padStart(9) + 'share'.padStart(8)
  );
  const ranked = [...samplesByName.entries()].sort((a, b) => b[1] - a[1]);
  for (const [name, count] of ranked.slice(0, kTopSymbols)) {
    const share = ((100 * count) / total).toFixed(1) + '%';
    console.log(
      '  ' +
        name.slice(0, 46).padEnd(46) +
        String(count).padStart(9) +
        share.padStart(8)
    );
  }
  console.log(`\n  attributed samples: ${total}`);
}

if (require.main === module) {
  const args = process.argv.slice(2);
  if (args.length !== 2) {
    console.error('Usage: symbolize.js IMAGE.elf capture.txt');
    process.exit(2);
  }
  const [elfFilePath, captureFilePath] = args;

  (async () => {
    try {
      await symbolize(elfFilePath, captureFilePath);
    } catch (e) {
      console.error(e.message);
      process.exit(1);
    }
  })();
}
